### PR TITLE
[feat.angles] new property angle_indexes

### PR DESCRIPTION
**Pull request by gph82**
This is a suggestion as to how feat.angle_indexes should work if the feature was called with cossin=True. This way, the column/dimension indices of 
- feat.get_output(), 
- feat.describe()
- feat.angle_indexes
  all match.

Internally, all computations are done with feat._angle_indexes as they were before. For the new behavior:

``` python
feat = pyemma.coordinates.featurizer(pdbfile)
feat.add_angles([[0,1,2], [2,3,4]] )
feat.add_angles([[0,1,2], [2,3,4]], cossin=True)
feat.add_dihedrals([[0,1,2,4], [4,5,7,8]])
feat.add_dihedrals([[0,1,2,4], [4,5,7,8]], cossin=True)

for act_feat in feat.active_features:
    print("act_feat.dimension= %u, cossin=%s"%(act_feat.dimension, act_feat.cossin))
    print("len(act_feat.describe) = %u"%len(act_feat.describe()))
    for l in act_feat.describe():
        print(l)
    print("act_feat.angle_indexes = ")
    print(act_feat.angle_indexes)
    print()
```

yielding

```
act_feat.dimension= 2, cossin=False
len(act_feat.describe) = 2
ANGLE: ARG 1 N 0 - ARG 1 CA 1 - ARG 1 C 2 
ANGLE: ARG 1 C 2 - ARG 1 O 3 - ARG 1 CB 4 
act_feat.angle_indexes = 
[[0 1 2]
 [2 3 4]]

act_feat.dimension= 4, cossin=True
len(act_feat.describe) = 4
ANGLE: COS(ARG 1 N 0 - ARG 1 CA 1 - ARG 1 C 2)
ANGLE: SIN(ARG 1 N 0 - ARG 1 CA 1 - ARG 1 C 2)
ANGLE: COS(ARG 1 C 2 - ARG 1 O 3 - ARG 1 CB 4)
ANGLE: SIN(ARG 1 C 2 - ARG 1 O 3 - ARG 1 CB 4)
act_feat.angle_indexes = 
[[0 1 2]
 [0 1 2]
 [2 3 4]
 [2 3 4]]

act_feat.dimension= 2, cossin=False
len(act_feat.describe) = 2
DIH: ARG 1 N 0 - ARG 1 CA 1 - ARG 1 C 2 - ARG 1 CB 4 
DIH: ARG 1 CB 4 - ARG 1 CG 5 - ARG 1 NE 7 - ARG 1 CZ 8 
act_feat.angle_indexes = 
[[0 1 2 4]
 [4 5 7 8]]

act_feat.dimension= 4, cossin=True
len(act_feat.describe) = 4
DIH: COS(ARG 1 N 0 -  ARG 1 CA 1 - ARG 1 C 2 - ARG 1 CB 4)
DIH: SIN(ARG 1 N 0 -  ARG 1 CA 1 - ARG 1 C 2 - ARG 1 CB 4)
DIH: COS(ARG 1 CB 4 -  ARG 1 CG 5 - ARG 1 NE 7 - ARG 1 CZ 8)
DIH: SIN(ARG 1 CB 4 -  ARG 1 CG 5 - ARG 1 NE 7 - ARG 1 CZ 8)
act_feat.angle_indexes = 
[[0 1 2 4]
 [0 1 2 4]
 [4 5 7 8]
 [4 5 7 8]]
```

Works with all angle features

### CODE REVIEW

File: pyemma/coordinates/data/featurization/angles.py [hunk 1]
```diff
@@ -75,13 +75,20 @@ def transform(self, traj):
...
         hash_value ^= hash_top(self.top)
         hash_value ^= hash(self.deg)
         hash_value ^= hash(self.cossin)
 
         return hash_value
 
+    @property
+    def angle_indexes(self):
+        if self.cossin:
+            angle_indexes = np.vstack([np.vstack((ai, ai)) for ai in self._angle_indexes])
```
Comment (marscher) [inline pyemma/coordinates/data/featurization/angles.py]: np.repeat is meant for this purpose - I'd guess.

**Comment by gph82** [comment]
Thanks @marscher !

**Comment by franknoe** [comment]
Not sure if I want to merge this. I understand angle_indices in conjunction with cossin=True as an input parameter which defines which features to construct. Why should it be accessible later, and why should it match the features in length? These are two different things.

**Comment by gph82** [comment]
So, the idea is that you might want to gain access to the indices behind, say, the 3rd feature. Not as strings (like in af.describe(), but as real indices that you can match to atoms, say in a vmd visualization,  If you did cossin=True, and lookup the 3rd entry in indices...you're looking at the wrong quad of integers. It is a suggestion. 

Another usecase is when down the analysis workflow you do TICA and want to know what dihedral (=what atoms) are most correlated with, say, the first TIC. So you use tica.feature_TIC_correlation, it comes out that the 11-feature is correlated to .99 percent...looking at angle_indexes[11] is wrong, you would have to look at the np.floor(11/2)-entry of af.angle_indexes

**Comment by franknoe** [comment]
Am 23/03/16 um 14:15 schrieb Guillermo Pérez-Hernández:

> So, the idea is that you might want to gain access to the indices 
> behind, say, the 3rd feature. Not as strings (like in af.describe(), 
> but as real indices that you can match to atoms, say in a vmd 
> visualization, If you did cossin=True, and lookup the 3rd entry in 
> indices...you're looking at the wrong quad of integers. It is a 
> suggestion.
> 
> Another usecase is when down the analysis workflow you do TICA and 
> want to know what dihedral (=what atoms) are most correlated with, 
> say, the first TIC. So you use tica.feature_TIC_correlation, it comes 
> out that the 11-feature is correlated to .99 percent...looking at 
> angle_indexes[11] is wrong, you would have to look at the 
> np.floor(11/2)-entry of af.angle_indexes
> 
> Doesn't work in general, because if you combine multiple features this 
> will not be [11] anymore. This is why we have describe.
> A general way to solve this would be to encode the feature metadata in 
> an object that could hold. Such an object could also offer a way to 
> reencode the information to a VMD selection (or other selection language)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/pull/741#issuecomment-200339951

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by gph82** [comment]
yeah...guess I had not thought about the general case...guess it'd be nice to have such an object, connecting features (composite) and groups of atoms to visualize

**Pull request closed without merging**